urllib3/packages/selectors.py

```diff
@@ -0,0 +1,288 @@
+# Backport of selectors.py from Python 3.5+ to support Python < 3.4
+# Also has the behavior specified in PEP 475 which is to retry syscalls
+# in the case of an EINTR error. This module is required because selectors34
+# does not follow this behavior and instead returns that no dile descriptor
+# events have occurred rather than retry the syscall. The decision to drop
+# support for select.devpoll is made to maintain 100% test coverage.
+
+import errno
+import select
+import six
+from collections import namedtuple, Mapping
+
+import time
+if hasattr(time, "monotime"):
+    monotonic = time.monotonic
+else:
+    monotonic = time.time
+
+EVENT_READ = (1 << 0)
+EVENT_WRITE = (1 << 1)
+
+HAS_SELECT = True  # Variable that shows whether the platform has a selector.
+
+
+def _fileobj_to_fd(fileobj):
+    """ Return a file descriptor from a file object. If
+    given an integer will simply return that integer back. """
+    if isinstance(fileobj, six.integer_types):
+        fd = fileobj
+    else:
+        try:
+            fd = int(fileobj.fileno())
+        except (AttributeError, TypeError, ValueError):
+            raise ValueError("Invalid file object: {0!r}".format(fileobj))
+    if fd < 0:
+        raise ValueError("Invalid file descriptor: {0}".format(fd))
+    return fd
+
+
+def _syscall_wrapper(func, syscall_timeout, *args, **kwargs):
+    """ Wrapper function for syscalls that could fail due to EINTR.
+    All functions should be retried if there is time left in the timeout
+    in accordance with PEP 475. """
+    if syscall_timeout is None:
+        expires = None
+    else:
+        timeout = float(syscall_timeout)
+        if timeout < 0:
+            timeout = 0
+        expires = monotonic() + timeout
+
+    result = None
+    while expires is None:
+        current_time = monotonic()
+        if current_time > expires:
```

Comment:
With `timeout=0` this may fail immediately, because `expires` is calculated based off an earlier call to `monotonic()`. We should probably allow one run with `timeout=0`, so you may want to have `current_time` updated once before this loop begins (and use it to calculate `expires`), and then update at the *bottom* of the loop.